File: frontend/src/types/filtersConfig.ts
export type Filter = {
  label: string;
  value: string; // ??? how do we represent null values?
};

export type FilterCategory = {
  id: string;
  label: string;
  filters: Filter[];
};

export type FilterCategoriesMap = Record<string, FilterCategory>;

export type FilterCategoryGroup = {
  name: string;
  categories: string[]; // an array of filter category ids
};

export type FiltersView = {
  name: string;
  categoryGroups: FilterCategoryGroup[];
  otherCategoryGroups: FilterCategoryGroup[]; // order in the array will be used for display order
};

export type FiltersConfig = {
  filterCategories: FilterCategoriesMap;
  filterViews: FiltersView[];
};

File: backend/models/filters_config.py
from pydantic import BaseModel, Field, ConfigDict


class FilterOption(BaseModel):
    label: str
    value: str


class FilterCategory(BaseModel):
    # e.g. "genotype-genome", "phenotype-Antibiotic_abbreviation"
    id: str
    label: str
    # e.g. "genotype" | "phenotype"
    dataset: str
    filters: list[FilterOption]


class Column(BaseModel):
    # e.g. "phenotype-Antibiotic_name"
    id: str
    label: str
    sortable: bool
    rank: int
    enable_by_default: bool = Field(alias="enable_by_default")


class CategoryGroup(BaseModel):
    name: str
    # list of category ids (column_id strings), e.g. ["phenotype-Antibiotic_abbreviation"]
    categories: list[str]


class FilterView(BaseModel):
    urlName: str
    name: str
    categoryGroups: list[CategoryGroup]
    otherCategoryGroups: list[CategoryGroup]
    columns: list[Column]


class FiltersConfig(BaseModel):
    model_config = ConfigDict(validate_by_name=True)

    filter_categories: dict[str, FilterCategory] = Field(alias="filterCategories")
    filter_views: list[FilterView] = Field(alias="filterViews")
    release: dict = Field()
